### PR TITLE
Replace CatalogEntityDetailRegistry with reactive solution

### LINKED ISSUE #4819: CatalogEntityDetailRegistry

### PR DESCRIPTION
**Pull request by jansav**
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

Closes #4819

### CODE REVIEW

File: src/renderer/components/+catalog/catalog-entity-details.tsx
Comment (Nokel81) [inline src/renderer/components/+catalog/catalog-entity-details.tsx]: nit: does this really need to be spread out over so many lines?

File: src/renderer/components/+catalog/details-for-catalog-entity.injectable.ts
Comment (Nokel81) [inline src/renderer/components/+catalog/details-for-catalog-entity.injectable.ts]: Do we really want to increase our dependence on lodash?

File: src/renderer/components/+catalog/detail-item-registrations.injectable.ts
Comment (Nokel81) [inline src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: This is something we do a lot within the extension API, can we try and factor it out into a common utility?

Comment (jansav) [reply src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: Nothing to be shared here. `orderBy` would be it, but it's already reusable. Extracted to separate function just to make use-place cleaner.

Comment (Nokel81) [reply src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: No it is shared, we have a bunch of types that extend `{ priority?: number }`, so I could imagine the following function

```ts
interface PrioritySortable {
  priority?: number;
}

export function orderByPriority<T extends PrioritySortable>(objs: T[]) {
  const withDefaultPriority = objs.map(({ priority = 50, ...rest }) => ({ priority, ...rest }));

  return orderBy("priority", "desc", withDefaultPriority);
}
```

File: src/renderer/components/+catalog/details-for-catalog-entity.injectable.ts
Comment (jansav) [reply src/renderer/components/+catalog/details-for-catalog-entity.injectable.ts]: TBH `lodash` is only library that I think we should depend without creating abstractions. That's because it provides helper functions which can be replaced relatively easily if ever needed. `lodash` provides a lot simplicity and ability to get rid nested code when combined with functional programming. (which we don't do at the moment)

File: src/renderer/components/+catalog/detail-item-registrations.injectable.ts
Comment (jansav) [reply src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: Sure, as long as there is no other usages which require default priority other than `50`. 

And when we start sharing functions it becomes even more important that function does only one thing and name doesn't lie. In example you provided function lies and does two things.

1. Adds default priority
2. Orders by priority

Just to keep in mind in future, since it feels acceptable in this case.

I'll do the change.

Comment (jansav) [reply src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: Done. Keep in mind that base of this branch changed. Do not merge until #4817 has been merged.

File: src/renderer/utils/order-by-priority.ts
Comment (ixrock) [inline src/renderer/utils/order-by-priority.ts]: Maybe this file should be located in `common/utils`?

File: src/renderer/components/+catalog/detail-item-registrations.injectable.ts
Comment (ixrock) [inline src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: When first time see this code/file it's hard to predict what kind of type it returns/accepts? (without extra navigating in codebase).
Maybe we should start using definition with type? Kinda `const detailItemRegistrationsInjectable: Injectable<string[]> = ...` so `detailItemRegistrations` expected to be strings array. Wdyt?

File: src/renderer/components/+catalog/details-for-catalog-entity.injectable.ts
Comment (ixrock) [inline src/renderer/components/+catalog/details-for-catalog-entity.injectable.ts]: Random question: if i replace lifecycle to `lifecycleEnum.singleton` how it would it break this injectable usages?

**Review by ixrock — DISMISSED** [review]

File: src/renderer/components/+catalog/details-for-catalog-entity.injectable.ts
Comment (jansav) [reply src/renderer/components/+catalog/details-for-catalog-entity.injectable.ts]: Two things:

1. Injectable would throw because we would be using `singleton` with `instantiationParameter`. (see below)
2. (if 1. defensive code would not be present) CatalogEntity used would be same for every time when injected.

However, this is something that we are actually working on, because we see this bad design. 

This should be `singleton` but only in scope of `instantiationParameter`. So if you give `"some-string"` and `"some-other-string"` as `instantiationParameter`, you will get two different instances, but if you inject `"some-string"` twice, you will get same instance. This allows us to get rid of these type of transient lifecycle usages and start using `computed` as a mechanism for achieving reactive UI.

File: src/renderer/components/+catalog/detail-item-registrations.injectable.ts
Comment (jansav) [reply src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: Hm. I think it would help a lot if I just extracted implementation to separate function.

File: src/renderer/utils/order-by-priority.ts
Comment (jansav) [reply src/renderer/utils/order-by-priority.ts]: Hm, maybe. It's used only in renderer though and that's why it's there.

File: src/renderer/components/+catalog/detail-item-registrations.injectable.ts
Comment (Nokel81) [reply src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: Why was the branch changed if that PR doesn't also use the function?

Comment (jansav) [reply src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: This PR makes the change for both.

Comment (Nokel81) [reply src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: That is weird, why have it so that we merge a PR and then immediately merge another PR that changes it?

Comment (jansav) [reply src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: Because both of these need the same code. One has to be merged first so other can use it. I don’t think PR process should cripple refactoring. 

End result is same and both of these are relatively small PRs.

Comment (Nokel81) [reply src/renderer/components/+catalog/detail-item-registrations.injectable.ts]: I would have just thought that adding the function in that PR and then just using it in this PR would make more sense and have a cleaner history.

File: src/renderer/utils/order-by-priority.ts
Comment (panuhorsmalahti) [reply src/renderer/utils/order-by-priority.ts]: I think code that can be used in both should be placed in common.

**Comment by jansav** [comment]
@panuhorsmalahti @Nokel81 @ixrock: This branch has been completely reworked with modern tools. PTAL again.

File: src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts
Comment (Nokel81) [inline src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:30]: is the priority negative?

File: src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts
Comment (Nokel81) [inline src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:38]: Why is this better than using `computed` directly? Now we have a whole bunch of instantiates every time any single extension gets added or removed or adds an item.

File: src/renderer/components/+catalog/catalog-entity-detail-items/catalog-entity-detail-item-injection-token.ts
Comment (Nokel81) [inline src/renderer/components/+catalog/catalog-entity-detail-items/catalog-entity-detail-item-injection-token.ts:20]: Why is this optional?

Comment (jansav) [reply src/renderer/components/+catalog/catalog-entity-detail-items/catalog-entity-detail-item-injection-token.ts:20]: It's optional because there is no need to limit `catalog-entity-detail-item` for extensions. Same interface is used for detail-items that live in OpenLens. Notice that this relates to other comment.

File: src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts
Comment (jansav) [reply src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:30]: Because we changed it to `orderNumber`. Priority has double meaning; it doesn't tell whether items should be sorted in ascending or descending order. `orderNumber` does so. It also allows us to use `sortBy` from `lodash/fp`

File: src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts
Comment (jansav) [reply src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:38]: `computed` has limitations. We noticed those during refactoring of routes. `computed` is especially problematic when there is `React` components inside. It's very easy to create new instance of a component every time.

`registrator`-pattern is something that we are trying for couple reasons: 
1. Extensions can just register more implementations of an injection token (use places don't need to prepare for it)
2. We have full control of given registrator in single place. Which means that we can keep use places flexible and create adapters for Extension API whenever needed without introducing breaking changes.

There is still some downsides in `registator`-pattern, most critical ones are the usage of installation counter and that we need to filter results of `injectMany` based on enabled extensions. This is however something that we are actively solving and the end goal is that you only need to `injectMany` and get results as reactive.

Comment (jansav) [reply src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:38]: Note: This is still something that we are trying out to see if it carries us over the finish line or not.

File: src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts
Comment (Nokel81) [reply src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:30]: But then we have to remember to make them negative, could sort by increasing by default?

File: src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts
Comment (Nokel81) [reply src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:38]: > computed is especially problematic when there is React components inside. It's very easy to create new instance of a component every time.

What do you mean by this? Wouldn't some form of shallow equality or reference marker be the solution then?

Comment (jansav) [reply src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:38]: For example: 

In case of pages we have logic that makes page component observable if page params are defined. It also provides page params as props for the component. This requires that we introduce higher-order component for Page component if page params are present. This higher-order component can't be created inside `computed` since that would cause React to treat it as a new component (re-mount) every time that the value of computed recomputes.

```typescript
computed(() => {
  return pipeline({
    extensions,
    flatMap(extension => extension.clusterPages),

    map(pageRegistration => ({
      ...pageRegistration,
      
      components: {
        ...pageRegistration.components,
        
        ...(pageRegistration.params ? {
          Page: observer(() => (<pageRegistration.components.Page />))
        } : {})
      }
    });
  });
});
```

Notice that I don't remember the exact PageRegistration API so property names could be wrong.

File: src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts
Comment (jansav) [reply src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:30]: No need, in extension API property name is still `priority` and it's positive, but in OpenLens (and hopefully in Extension API 2.0) property name is `orderNumber` which is also positive. Code that we are discussing about is the adapter which converts Extension API 1.0 to internal API.

File: src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts
Comment (Nokel81) [reply src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:38]: Why do we need to do that? I would argue that since the page dereferences computed values, if it is not marked as observable itself, that is a bug or at least wrong and should be fixed. Locality of use is key here.

Comment (jansav) [reply src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:38]: I completely agree with you on that. We shouldn't make the component `observer` in behind the scenes. However, that's how current implementation works and is exposed in Extension API so we can't make the change without introducing breaking change.  

Anyway, that's just one requirement. There might be other requirements for higher-order components which cause the same exact issue. E.g. we add `TabLayout` higher order component to page components if they have siblings.

Comment (jansav) [reply src/renderer/components/+catalog/catalog-entity-detail-items/extension-catalog-entity-detail-items-registrator.injectable.ts:38]: We are currently testing the hypothesis that Extension register new implementations of an `injectionToken` which then allows us to treat extensions as first-class citizens and therefore introduce a lot simplicity. Imagine that everything is built in a way that if we that we want to expose something to Extension API, we just expose the interface for injection token. In that world, we have:

1. One place that registers all implementations of `injectionToken` from extension (extension is enabled)
2. One place that deregisters all injectables that was registered because of extension was enabled (extension is disabled)

In that way we can use `di.injectMany` (or something similar) safely everywhere we want to have all implementations and we don't need to worry about enabled extensions etc. Also `registrator` will be consolidated to one place. 

Treat `registrator` as an adapter for Extension API 1.0 -> Extension API 2.0, which allows us to experiment with requirements of an interface before we actually expose the interface to extensions. At the same time it allows us to start working on the changes required (if any) in use places of said implementations.

**Pull request closed without merging**
